### PR TITLE
Constrain priors with symmetric mass distribution

### PR DESCRIPTION
**Pull request by lucianopaz**
**What is this PR about?**
The `find_constrained_prior` function was optimising the difference between the upper and lower cdf values so that it matched a particular mass. This can be very brittle since there are multiple lower and upper cdf values that produce exactly the same mass. For example, [0.01, 0.96] will have 0.95 mass but [0.025, 0.975] will have it too. This PR proposes to use a different optimisation target so that the prior will be constrained to have the desired mass, and that the lower and upper cdf bounds will be symmetric.

**Checklist**
+ [X] Explain important implementation details 👆
+ [ ] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [ ] Are the changes covered by tests and docstrings?
+ [X] Fill out the short summary sections 👇

## Bugfixes / New features
- find_constrained_prior now attempts to find the parameters that lead to the desired mass between the lower and upper bounds, but also forces the lower and upper cdf values to be symmetric.

### CODE REVIEW

**Review by ricardoV94 — COMMENTED** [review]
I think the reason for this rule was for positive distributions, so that you can specify lower=0, upper=x? I don't remember exactly.

If this or something similar we might want to allow both symmetric and assymetric rules.

**Comment by ricardoV94** [comment]
Can you share the flaky example? I would imagine that with the 3 constraints (lower, upper, and mass in between), most distributions of up tp 3 parameters would be uniquely constrained

**Comment by lucianopaz** [comment]
> I think the reason for this rule was for positive distributions, so that you can specify lower=0, upper=x? I don't remember exactly.
> 
> If this or something similar we might want to allow both symmetric and assymetric rules.

But with the previous rule, the optimisation target has no way of knowing that you want lower=0 and upper=mass. So it will still be underdetermined.

**Comment by lucianopaz** [comment]
> Can you share the flaky example? I would imagine that with the 3 constraints (lower, upper, and mass in between), most distributions of up tp 3 parameters would be uniquely constrained

```python
pm.find_constrained_prior(
    pm.Gamma,
    lower=20,
    upper=700,
    mass=0.95,
    init_guess=dict(alpha=5.29, beta=0.023),
)
```

If I’m not missing anything, I’m just constraining lower and upper. The mass being in the lower upper interval is a side effect.

**Comment by ricardoV94** [comment]
> If I’m not missing anything, I’m just constraining lower and upper. The mass being in the lower upper interval is a side effect.

Lower and upper are constants and, given those, it's optimizing the mass in between. 

Okay, so the problem (as you wrote) is that you can just shift the distribution left and right and keep the interval mass between upper and lower constant right?

Should we add a second constraint for the mass to the left of the first point which defaults to `(1-mass)/2` (your PR) but can also be set to (eg, 0) if you want to set an absolute lower bound for positive variables? 

I have the feeling that with positive distributions I would often want to think about stuff in the range (0-x) but maybe it's just me, or it makes things unnecessarily complicated.

**Comment by lucianopaz** [comment]
> Should we add a second constraint for the mass to the left of the first point which defaults to `(1-mass)/2` (your PR) but can also be set to (eg, 0) if you want to set an absolute lower bound for positive variables?

Yeah, I can add that.

> I have the feeling that with positive distributions I would often want to think about stuff in the range (0-x) but maybe it's just me, or it makes things unnecessarily complicated.

Maybe when you’re working with an exponential distribution the [0, mass] range is more natural, but not for the gamma or the lognormal distributions. Anyway, we can include the option to set the lower bound.

**Comment by aloctavodia** [comment]
This is one of the reasons that in PreliZ we use the maximum entropy constraint. https://preliz.readthedocs.io/en/latest/api_reference.html#preliz.maxent. My plan is to have a usable version of PreliZ this year.

**Comment by lucianopaz** [comment]
> This is one of the reasons that in PreliZ we use the maximum entropy constraint. https://preliz.readthedocs.io/en/latest/api_reference.html#preliz.maxent. My plan is to have a usable version of PreliZ this year.

Yeah, entropy is better behaved than the `exp(logcdf)` at some edge. @aloctavodia, would there be any benefit of implementing the entropy of the distributions? Maybe it could be useful for aeppl?

**Comment by aloctavodia** [comment]
Not sure if having the entropy of the distributions will be of general use. At this point I can only think of maximum entropy priors. And in a few months I think we will be ready to deprecate find_constrained_prior in favor of PreliZ.

File: pymc/func_utils.py
Comment (AlexAndorra) [inline pymc/func_utils.py]:
```suggestion
        mass below the ``lower`` value will be equal to the probability
```

File: pymc/func_utils.py
Comment (AlexAndorra) [inline pymc/func_utils.py:166]: Are we sure we don't need this check anymore @lucianopaz ?

File: pymc/func_utils.py
Comment (AlexAndorra) [inline pymc/func_utils.py:67]: I would add a small example of why that would be useful

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Looks better indeed, thanks @lucianopaz ! Just had a few comments before merging

File: pymc/func_utils.py
Comment (lucianopaz) [reply pymc/func_utils.py:166]: Yes, we don't need it anymore. Now that the mass condition is an optimization constraint, the optimization will fail if the condition is not met.

File: pymc/func_utils.py
Comment (ricardoV94) [inline pymc/func_utils.py]:
```suggestion
    Under some circumstances, you might not want to have the same cumulative
```

**Review by ricardoV94 — APPROVED** [review]
Looks great!

**Review by AlexAndorra — APPROVED** [review]
All good now, thanks @lucianopaz ! I'll merge once tests pass

**Pull request merged**
